### DISCUSSION TITLE
Valtio with persisted state

### DISCUSSION THREAD

**Discussion by GollyJer** [pmndrs/valtio#319] [Show and tell]
OK, I put about 20 hours into playing with and understanding Valtio over the last few days. Here's a little story and then a codesandbox with what I've come up with. Hopefully it helps others. Or @dai-shi can tell me I'm doing it wrong. 😀

We have a React Native app, [token.art NFT viewer](https://token.art/), that makes extensive use of "computed" state and persistence. Our current state package isn't being maintained and has some bugs. It is time to make the switch.

Here are the requirements.

- Supports Typescript types. Ideally written in Typescript.
- Rehydration and data migrations can happen before `<App>` is rendered.
- A [persistence package](https://github.com/Noitidart/valtio-persist) is available. ✅ Thanks @Noitidart. 🎉
- State slices can be persisted separately.
- State files can logically contain state & actions in one file.
- We can structure files so everything is super easy to consume in app; ideally one export file (stretch goal).

### [This codesandbox](https://codesandbox.io/s/react-valtio-with-persisted-state-v4yoy?file=/src/store/state/index.ts) shows what I came up with.

The key insight came when reading the wiki. Specifically, [How to Organize Actions](https://github.com/pmndrs/valtio/wiki/How-to-organize-actions). It gave me the inspiration to separate the exports.  State is a `default export` and actions are grouped together in a single `named export`. 

<img width=400 src=https://user-images.githubusercontent.com/689204/148841061-0ba82aef-a306-41c8-981a-531f171bab5a.png>

This allowed the passing of only the state object to valtio-persist... 

<img width=600 src=https://user-images.githubusercontent.com/689204/148843602-049b87e5-1621-4856-b27e-7a92e3ace999.png>

...and re-importing the persisted state proxy back into the "state" file for use in actions. 
See (`import { state } from "./";` in thing.ts above).

Then a nested folder structure with a well placed index file that re-exports all the named exports...

<img width=600 src=https://user-images.githubusercontent.com/689204/148842073-121f11f7-539c-4245-8e85-fdce6601cc47.png>

... leads to persisted and reactive state that is easy to access from the top level "store" import.

<img width=600 src=https://user-images.githubusercontent.com/689204/148839900-c9867822-d152-4908-a461-e1f35efd2e1c.png>

And last but not least... types are correctly inferred and used for auto-complete and visible on hover.
<img width=600 src=https://user-images.githubusercontent.com/689204/148842962-ffab6569-1ffd-4598-b875-c31a6e56ddb0.png>

Now adding a new piece of persisted state is as simple as creating a new file in the state folder, adding the default import to persist, and reexporting the actions at the store level. 
🎉

### [Here's the codesandbox](https://codesandbox.io/s/react-valtio-with-persisted-state-v4yoy?file=/src/store/state/index.ts) again. Playing with it will help.

I'm very curious to hear anyone's opinions. 
For now this is the approach I'm going to use in our app. 👍

**Comment by dai-shi** [comment]
Nicely done.
I don't think you need to export `thing1` and re-import from parent. (if I understand how `proxyWithPersist` works correctly.)
Please git it a try.

```js
const thing1 = proxy(INITIAL_STATE)
const add = () => { thing1.number += 1 }
export default thing1
export const thing1Actions = { add }
```

**Reply by GollyJer** [reply]
Thanks @dai-shi. 👍

Yeah. That does work but I was running into one major issue with it; [like you discussed here](https://github.com/pmndrs/valtio/discussions/199#discussioncomment-1093597). Resetting of initial state is more complicated when the variable references a proxy directly.

We end up with `Cannot assign to 'thing1' because it is a constant.`.
<img width=300 src=https://user-images.githubusercontent.com/689204/148864785-2517813b-6dd4-4bcc-a74f-334b2a56d676.png>

Have you found a better pattern for `reset`?

**Reply by dai-shi** [reply]
I would do this:
```js
const reset = () => {
  Object.keys(INITIAL_STATE).forEach((key) => {
    thing1[key] = INITIAL_STATE[key];
  });
};
```
It's not TypeScript friendly, but we know what we are doing.
Note: It will not delete keys that are not in `INITIAL_STATE`.

**Reply by GollyJer** [reply]
Ah! I was trying, and failing, to loop through and delete each key first and then add back initial_state.
I think this combination of our ideas gets us there.

```javascript 
const reset = () => {
  Object.keys(thing1).forEach((key) => {
    if (INITIAL_STATE[key]) {
      thing1[key] = INITIAL_STATE[key];
    } else {
      delete thing1[key];
    }
  });
};
```

I don't love Typescript screaming at me but doing it this way eliminates the need for the re-import which I think is worth it. I'll update the codesandbox used in the original post.

**Reply by dai-shi** [reply]
> doing it this way eliminates the need for the re-import which I think is worth it.

I agree with it.

`if (INITIAL_STATE[key]) {` this affects all falsy values like `0`, if you want to be strict about it 👉  `if (key in INITIAL_STATE) {`.

**Reply by GollyJer** [reply]
Good catch. Here's an updated version. 
Added types. Can't stand those red squiggles. 🙂

```javascript 
type InitialStateType = Partial<typeof INITIAL_STATE>;
const reset = () => {
  const objToReset = thing1 as InitialStateType;
  let key: keyof InitialStateType;

  for (key in objToReset) {
    if (key in INITIAL_STATE) {
      objToReset[key] = INITIAL_STATE[key];
    } else {
      delete objToReset[key];
    }
  }
};
```

**Reply by dai-shi** [reply]
another nit: if some code did `delete thing1.foo`, it will be not reset even if `INITIAL_STATE` has a `foo` property.

**Reply by GollyJer** [reply]
Another good catch. What about this one?
Let's just clear the whole thing out and rebuild it.
```javascript 
type InitialStateType = Partial<typeof INITIAL_STATE>;

const reset = () => {
  const objToReset = thing1 as InitialStateType;
  let key: keyof InitialStateType;

  for (key in objToReset) {
    delete objToReset[key];
  }

  for (key in INITIAL_STATE) {
    objToReset[key] = INITIAL_STATE[key];
  }
};
```

**Reply by dai-shi** [reply]
looks clean

**Reply by GollyJer** [reply]
It turns out that this approach (the original topic of this discussion) doesn't work, at all, with react-native and async storage. 😭

**Reply by Noitidart** [reply]
Oh no, why is that? I'm using it successfully with Expo. What part is not working?

**Reply by GollyJer** [reply]
I'll hit you up in your github about the issues. My app is also built with Expo. A major issue I ran into was using an array in the proxy.

Also had to make my own `storage` to get it working and had to make a lot of changes to persist so I pulled in your raw code and started hacking away. 

Let's definitely collaborate on making a great `valtio-persist` package. 
Finishing up an app release today and then I'll post in your issues. 👍

**Reply by Noitidart** [reply]
Oh thank you! Oh an array use case that's very interesting. I think valtio-persist should definitely support that. I'll work on something right now.

**Comment by Noitidart** [comment]
Thanks for the call out and trying out the lib. Way appreciate it!

> State slices can be persisted separately.

That was a big one for me too, so I included that in the lib. :) 

I do have to add error handling, like when a file read fails it should log or something. redux-persist was also silent about errors, so I didn't prioritize this in v1.

**Reply by GollyJer** [reply]
Yeah man. We have one object that is huge (and not written often) and the rest are pretty small. Re-writing all of them each time was such a drain. `redux-persist` didn't have this option and stumbling on valtio AND your package that came along with it. Pure gold! Thanks again.

I did have one specific question I wasn't sure about. Might have slipped through the notification cracks. Pretty sure it has to do with serializing via JSON.stringify.
👉 https://github.com/Noitidart/valtio-persist/issues/1

**Reply by Noitidart** [reply]
Oh shoot so sorry, checking now, missed that notification, thanks!

**Comment by Noitidart** [comment]
For this image below, I see default export is `proxy(initialState)`. Why is this done? Why not just export the initialState without wrapping it into a proxy?

The one reason I could see on wrapping initialState in proxy in that file, is if you want to make your actions work on that. So you don't have to re-import the persisted state.

So the options I'm thinking of up there are this:

### Option 1 - export initial state and re-import persisted state

```
import { state } './';

export default initialState;

const add = () => {
  state.thing1.number = 1;
}
```

### Option 2 - export proxied initial state, and connect actions to this proxy. no re-import of persisted state

I'm not sure if this actually works. But the idea here is to not re-import persisted state. Changes on this proxy will make the persisted proxy persist it.

```
const thing1 = proxy(initialState);

const add = () => {
  thing1.number = 1;
}

export default thing1;
```

and persisted:

```
import thing1 from './thing1';

const state = proxyWithPersist({
  initialState: {
    thing1
  },
  ...
});
```

![](https://user-images.githubusercontent.com/689204/148841061-0ba82aef-a306-41c8-981a-531f171bab5a.png)

**Reply by Noitidart** [reply]
oh please ignore, i think thats what daishi is mentioning above. and you mentioned reseting state was causing issue.

**Reply by GollyJer** [reply]
Thanks for the response. We're all on the same page. An ideal solution allows confidently resetting state without the need to reimport and do the object spread. We're working through a solution in the comments of Daishi's answer. 👍
